cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,793 @@
+                   hash_state.buffer + Hasher::message_chunk_size,
+                   0x00);
+      this->underlying().hash_step(hash_state);
+
+      // Fill the entire message with zeros up to the final bytes reserved for
+      // the message length.
+      thrust::fill_n(thrust::seq,
+                     hash_state.buffer,
+                     Hasher::message_chunk_size - message_length_supported_size,
+                     0x00);
+    }
+
+    // Convert the 64-bit message length from little-endian to big-endian.
+    uint64_t const full_length_flipped = swap_endian(message_length_in_bits);
+    memcpy(hash_state.buffer + Hasher::message_chunk_size - message_length_supported_size,
+           reinterpret_cast<uint8_t const*>(&full_length_flipped),
+           message_length_supported_size);
+    this->underlying().hash_step(hash_state);
+
+    // Each byte in the word generates two bytes in the hexadecimal string digest.
+    // SHA-224 and SHA-384 digests are truncated because their digest does not
+    // include all of the hash values.
+    auto constexpr num_words_to_copy =
+      Hasher::digest_size / (2 * sizeof(typename Hasher::sha_word_type));
+    for (int i = 0; i < num_words_to_copy; i++) {
+      // Convert word representation from big-endian to little-endian.
+      typename Hasher::sha_word_type flipped = swap_endian(hash_state.hash_value[i]);
+      if constexpr (std::is_same_v<typename Hasher::sha_word_type, uint32_t>) {
+        uint32ToLowercaseHexString(flipped, result_location + (8 * i));
+      } else if constexpr (std::is_same_v<typename Hasher::sha_word_type, uint64_t>) {
+        uint32_t low_bits = static_cast<uint32_t>(flipped);
+        uint32ToLowercaseHexString(low_bits, result_location + (16 * i));
+        uint32_t high_bits = static_cast<uint32_t>(flipped >> 32);
+        uint32ToLowercaseHexString(high_bits, result_location + (16 * i) + 8);
+      } else {
+        cudf_assert(false && "Unsupported SHA word type.");
+      }
+    }
+  };
+
+  template <typename T,
+            typename Hasher                                             = HasherT,
+            typename std::enable_if_t<(!is_fixed_width<T>() || is_chrono<T>()) &&
+                                      !std::is_same_v<T, string_view>>* = nullptr>
+  void CUDA_DEVICE_CALLABLE operator()(column_device_view,
+                                       size_type,
+                                       typename Hasher::sha_intermediate_data&)
+  {
+    cudf_assert(false && "SHA unsupported type.");
+  }
+
+  template <typename T,
+            typename Hasher                                    = HasherT,
+            typename std::enable_if_t<is_floating_point<T>()>* = nullptr>
+  void CUDA_DEVICE_CALLABLE operator()(column_device_view col,
+                                       size_type row_index,
+                                       typename Hasher::sha_intermediate_data& hash_state)
+  {
+    T const& key = col.element<T>(row_index);
```

Comment:
This function normalizes `-0.0` to hash the same as `0.0` (and similarly for `nan`). The two values are considered equal 
 (`0.0 == -0.0`) but the binary representation is not equal, meaning that their hashes would not be equal without this normalization. I am copying this behavior from the MD5 hash, but I am not sure if this is desired. Can a reviewer offer insight on whether we should prefer "bitwise hashing" (no normalization, hash the raw binary data) or "equality hashing" (`x == y` implies `hash(x) == hash(y)`)?

Reply:
Spark considers `0.0` and `-0.0` different for hashing purposes. See #7672. Though this only affects Murmur3 (MD5 and SHA can only be applied to bytes-like objects), this makes me lean more strongly towards favoring the bitwise approach for both MD5 and SHA in libcudf.